File: src/main/java/com/trilead/ssh2/crypto/keys/Ed25519Provider.java
package com.trilead.ssh2.crypto.keys;

import java.security.AccessController;
import java.security.PrivilegedAction;
import java.security.Provider;
import java.security.Security;

/**
 * JCA Provider for Ed25519 cryptographic operations.
 * <p>
 * Registers KeyFactory and KeyPairGenerator implementations for the
 * Ed25519 signature algorithm with the Java security framework.
 *
 * @see Ed25519KeyFactory
 * @see Ed25519KeyPairGenerator
 */
public class Ed25519Provider extends Provider {
	public static final String NAME = "ConnectBot Ed25519 Provider";
	public static final String KEY_ALGORITHM = "Ed25519";
	private static final Object sInitLock = new Object();
	private static boolean sInitialized = false;

	public Ed25519Provider() {
		super(NAME, "1.0", "Not for use elsewhere");
		AccessController.doPrivileged((PrivilegedAction<Object>) () -> {
			setup();
			return null;
		});
	}

	protected void setup() {
		put("KeyFactory." + KEY_ALGORITHM, getClass().getPackage().getName() + ".Ed25519KeyFactory");
		put("KeyPairGenerator." + KEY_ALGORITHM, getClass().getPackage().getName() + ".Ed25519KeyPairGenerator");

		//   id-Ed25519   OBJECT IDENTIFIER ::= { 1 3 101 112 }
		put("Alg.Alias.KeyFactory.1.3.101.112", KEY_ALGORITHM);
		put("Alg.Alias.KeyFactory.EdDSA", KEY_ALGORITHM);
		put("Alg.Alias.KeyFactory.OID.1.3.101.112", KEY_ALGORITHM);
		put("Alg.Alias.KeyPairGenerator.1.3.101.112", KEY_ALGORITHM);
		put("Alg.Alias.KeyPairGenerator.EdDSA", KEY_ALGORITHM);
		put("Alg.Alias.KeyPairGenerator.OID.1.3.101.112", KEY_ALGORITHM);
	}

	public static void insertIfNeeded() {
		synchronized (sInitLock) {
			if (!sInitialized) {
				Security.insertProviderAt(new Ed25519Provider(), 1);
				sInitialized = true;
			}
		}
	}
}

File: sshlib/src/main/kotlin/org/connectbot/sshlib/crypto/ed25519/Ed25519Provider.kt
/*
 * ConnectBot SSH Library
 * Copyright 2025-2026 Kenny Root
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

@file:Suppress("DEPRECATION", "removal")

package org.connectbot.sshlib.crypto.ed25519

import java.security.AccessController
import java.security.PrivilegedAction
import java.security.Provider
import java.security.Security

// Android only has the Provider(String, double, String) constructor.
internal class Ed25519Provider : Provider(NAME, 1.0, "ConnectBot Ed25519 JCA Provider") {
    init {
        AccessController.doPrivileged(
            PrivilegedAction {
                setup()
                null
            },
        )
    }

    private fun setup() {
        val packageName = Ed25519Provider::class.java.`package`.name
        put("KeyFactory.$KEY_ALGORITHM", "$packageName.Ed25519KeyFactory")
        put("KeyPairGenerator.$KEY_ALGORITHM", "$packageName.Ed25519KeyPairGenerator")

        // OID aliases: id-Ed25519 OBJECT IDENTIFIER ::= { 1 3 101 112 }
        put("Alg.Alias.KeyFactory.1.3.101.112", KEY_ALGORITHM)
        put("Alg.Alias.KeyFactory.EdDSA", KEY_ALGORITHM)
        put("Alg.Alias.KeyFactory.OID.1.3.101.112", KEY_ALGORITHM)
        put("Alg.Alias.KeyPairGenerator.1.3.101.112", KEY_ALGORITHM)
        put("Alg.Alias.KeyPairGenerator.EdDSA", KEY_ALGORITHM)
        put("Alg.Alias.KeyPairGenerator.OID.1.3.101.112", KEY_ALGORITHM)
    }

    companion object {
        const val NAME = "ConnectBot Ed25519 Provider"
        const val KEY_ALGORITHM = "Ed25519"

        private val initLock = Any()
        private var initialized = false

        fun insertIfNeeded() {
            synchronized(initLock) {
                if (!initialized) {
                    Security.insertProviderAt(Ed25519Provider(), 1)
                    initialized = true
                }
            }
        }
    }
}
